Introduce per-function environment for Emit

gretay-js opened the pull request:
This PR is a follow-up of the discussion in PR https://github.com/ocaml/ocaml/pull/8928#issuecomment-530354066.

Currently, Emit uses several reference variables that are reset for each function it emits.  Some of these reference variables are accessed only for read (not write) between resets.

This PR collects all per-function reference variables into an environment and passes it as an argument to emit_instr and other emitter functions. I think it makes the code easier to understand and modify.

I've implemented the changes for amd64 only. I'd like to get some feedback on this proposal, and then I'll make similar changes in other backends.

This PR consists of two commits: the first one converts reference variables into mutable fields and the second removes "mutable" from these fields and adjusts the code.

There is no change to the generated assembly.  I checked it by comparing the generated assembly for all *.ml files in the compiler itself.

Note that one of the fields of the environment holds current Linear.fundecl.  It eliminates the need for reference variables that were basically read-only copies of various field of fundecl. For example, in amd64 backend, it eliminates function_name, fastcode_flag, tailrec_entry_point, as well as the most-recently added prologue_required and frame_required. 

I tried to keep other changes in emit.mlp to minimum, but I had to move type declaration of gc_call and bound_error_call away from the functions that use them. The reason is that type declaration of env uses them and the functions that manipulate them take env as an argument.

The remaining reference variables are per compilation unit.

gretay-js committed:
Introduce per-function environment for Emit

gretay-js committed:
Remove 'mutable' from fields of Emit.env

lthls commented on asmcomp/amd64/emit.mlp:
```diff
@@ -66,35 +66,64 @@ let emit_debug_info dbg =
+  }
 
-let frame_required = ref false
+(* Tradeoff between code size and code speed *)
```
This comment applied to `fastcode_flag`, there's no reason to keep it.

lthls commented on asmcomp/amd64/emit.mlp:
```diff
@@ -251,53 +270,42 @@ let record_frame_label ?label live raise_ dbg =
-  I.add (int (index * 8)) (reg Proc.loc_spacetime_node_hole)
+  I.add (int (index * 8)) (reg Proc.loc_spacetime_node_hole ~env)
 
 (* Record calls to the GC -- we've moved them out of the way *)
```
You should move the comment to the `gc_call` type definition.

lthls commented on asmcomp/amd64/emit.mlp:
```diff
@@ -309,48 +317,39 @@ let emit_call_gc gc =
 (* Record calls to caml_ml_array_bound_error.
    In -g mode, or when using Spacetime profiling, we maintain one call to
    caml_ml_array_bound_error per bound check site.  Without -g, we can share
    a single call. *)
```
Same as above; this goes with `bound_error_call`

lthls commented on asmcomp/amd64/emit.mlp:
```diff
@@ -194,31 +223,21 @@ let reg64 = function
-
-let arg i n = reg i.arg.(n)
-
+let arg i n ~env = reg i.arg.(n) ~env
```
I'd prefer `reg ~env i.arg.(n)`, to match the order in `reg`'s definition

lthls commented on asmcomp/amd64/emit.mlp:
```diff
@@ -66,35 +66,64 @@ let emit_debug_info dbg =
+  mutable stack_offset : int;
+  mutable call_gc_sites : gc_call list;
+  mutable bound_error_sites : bound_error_call list;
+  mutable bound_error_call : int;
```
While we're at it, `bound_error_call` should be declared as `label`, and I would make it `label option` for clarity.

lthls reviewed (commented):
Great work, thank you very much !

I would be in favour of making `env` the first argument rather than the last in all functions, and I'm not sure the second commit is necessary (the traversal is one-way only, so it never needs to keep the old environment). The name `env` could be improved, too: I'd suggest calling the type something like `per_function_info` and the variables `info`, or `accumulators`/`acc` to reflect that its purpose is to hold and aggregate various kinds of data during emission of a single function.
I've also made some local comments on a few things that could be improved (mostly about comments).

I'd suggest waiting for other people to comment before changing things too much, my opinion may not be the most popular one. In particular, if you can keep the second commit around, that could help other reviewers assessing which one they prefer.

Octachron commented on asmcomp/amd64/emit.mlp:
```diff
@@ -66,35 +66,64 @@ let emit_debug_info dbg =
+    bound_error_sites = [];
+    bound_error_call = 0;
+    symbols_defined = String.Set.empty;
+    symbols_used = String.Set.empty;
```
The `symbols_defined` and `symbols_used` fields seem to not be used yet, is that an oversight?

lthls replied:
They're compilation unit-wide variables, so they should not be part of the function-specific environment. They're probably left-over from earlier attempts.

gretay-js replied:
ouch, yes, I forgot to remove them, sorry, and thanks for catching it!

gasche commented:
Re. naming, I don't like `acc` (too generic; I use it only for very local things, typically the "extra" argument of the tail-recursive version of a function.), and I find both `env` and `info` equally reasonable. For me "environment" clearly reasonates as "the thing you pass around everywhere to be able to hold the necessary information", and in particular the "Environment monad" is a common name for the monad capturing that pattern (also called "Reader Monad").

I like the idea of having a more specific/informative (and longer) name for the *type*, and keeping a concise name for the argument (if it occurs in every function declaration).

gretay-js committed:
Address comments about comments, names, and type of bound_error_call

gretay-js commented:
I'll change env to info and the order of arguments all over the place separately. I've addressed all the other comments.

Octachron commented:
I read the PR and I agree with @lthls , the change overall is nice. In particular, I agree that `env` should be made the first arguments of most functions. Similarly, I am mostly neutral on the second commit.

gretay-js committed:
Revert "Remove 'mutable' from fields of Emit.env"

This reverts commit 78cee96423ab3f41f145b92c82167e9d042fd2c8.

gretay-js committed:
Move env to be the first argument (amd64)

gretay-js committed:
Per function environment for i386

gretay-js committed:
Per function environment for arm

gretay-js committed:
Per function environment for arm64

gretay-js committed:
Branch_relaxation takes Linear.fundecl to compute instr_size

gretay-js committed:
Per function environment for power

gretay-js committed:
Per function environment for s390x

gretay-js committed:
Extract common part of per_function_env to Emitaux

gretay-js commented:
Thank you @lthls @Octachron for your reviews! I extended the PR to all targets.
Precheck ([build 317](https://ci.inria.fr/ocaml/job/precheck/317/)) is successful.

Other changes:
- Revert the second commit, leaving the fields of `per_function_env` as mutable.
- Change `env` to be the first argument everywhere and made it unnamed.
- Keep `env` variable name as is (rather than calling it `info`), based on the review feedback that env is also okay, now that we have a more descriptive type name `per_function_env` for it.

There is a lot of similarity in `per_function_env` for different targets.  @mshinwell suggested to extract the common definitions into Emitaux. The type of `per_function_env` is parametric to handle target-specific fields in arm* and power, and any future target-specific fields.

I wonder if there is a better way to implement the common per_function_env with target-specific extensions (using a functor? objects?).

In particular, it would be nice to improve on some of the following issues:
- The type of `gc_call` and `bound_error_call` are parametric only because on amd64 these record types have extra fields for spacetime.
- On power, call_gc_labels is implemented using a map and not a the common list as in all other targets.
- Code that accumulates pending literals, gc_calls and bound_error_calls can be shared, but doing so would involve some subtle changes that should be in a separately PR, I think.
- Some fields of per_function_env are used in some, but not all, targets, such as pending float_literals and included in the shared part.
- Target-specific fields of `per_function_env` require an extra deref.

gretay-js committed:
Update Changes

gretay-js commented:
I've rebased this PR.  
The removal of spacetime simplified this PR, in particular no need for amd64-specific fields in `gc_call` and `bound_error_call`.

I am getting a CI failure on i386-static that I cannot reproduce locally, but I have noticed similar failures in other PRs recently, so maybe the failure is unrelated?

Other than that, I think the PR is ready.
@lthls Could you have a look at this version of the code please?

lthls commented:
A small comment before I check the details: this PR introduces a `.ml` file without a corresponding `.mli` file (`emitenv_intf.ml`). This file only contains types, so it could be turned into a `.mli` file without much trouble. Do we have a policy on this ?

I'd also prefer not including the whole module in `Emitaux`. If all it takes is a few more `open Emitenv_intf`, then I'm fine with that (you could also remove the `_intf` part completely).

gretay-js committed:
Rename emitenv_intf.ml to emitenv.mli

gretay-js commented:
Ah, yes, there are several `mli` files without an `ml` for similar purposes, thanks!

I renamed `emitenv_intf.ml` to `emitenv.mli` and adjusted the code.

lthls commented on asmcomp/power/emit.mlp:
```diff
@@ -32,46 +33,55 @@ let reserved_stack_space =
   | ELF64v1 -> 48
   | ELF64v2 -> 32
 
-(* Layout of the stack.  The stack is kept 16-aligned. *)
```
It might make sense to keep this comment (putting it just before `initial_stack_offset`)

lthls commented on asmcomp/power/emit.mlp:
```diff
@@ -32,46 +33,55 @@ let reserved_stack_space =
-  | Incoming n -> frame_size() + reserved_stack_space + n
+  | Incoming n -> frame_size env + reserved_stack_space + n
+  (* CR: reserved_stack_space is already taken into account by frame_size.
+     Why is it added again when calculating slot_offset for Incoming? *)
```
I assume that it is because we're reaching into the caller's stack frame, so we need to take its reserved stack space into account too. Maybe @xavierleroy can confirm ?

gretay-js replied:
I've already asked it separately in https://github.com/ocaml/ocaml/issues/9084  and it's fine, I'll replace the CR with a comment as suggested in #9084.

lthls commented on asmcomp/power/emit.mlp:
```diff
@@ -1099,25 +1087,27 @@ let fundecl fundecl =
   end;
   (* Emit the jump tables *)
-  if !jumptables <> [] then begin
+  match env.p.jumptables_lbl with
```
You could match on `env.p.jumptables_lbl` and `env.p.jumptables` at the same time, which would feel more natural to me (and is slightly more efficient):
```
match env.p.jumptables, env.p.jumptables_lbl with
| _ :: _, None | [], Some _ -> assert false (* Sanity check *)
| [], None -> ()
| _ :: _, Some j -> ...
```

lthls commented on asmcomp/arm64/emit.mlp:
```diff
@@ -339,33 +307,27 @@ let emit_cmpimm rs n =
+let float_literal env fl =
   try
-    List.assoc f !float_literals
+    let x = List.find (fun x -> compare x.fl fl = 0) env.float_literals in
```
I believe that `compare` is better than `(=)` used by `List.assoc`, but I'm worried that it does not make a distinction between `0.` and `-0.` (and different NaN values, but that seems less problematic). Should this be fixed to use e.g. `Int64.equal (Int64.bits_of_float x.fl) (Int64.bits_of_float fl)` ? cc @xavierleroy

mshinwell replied:
Indeed, I think this should be using the comparison on the `Int64.bits_of_float` representation.

lthls commented on asmcomp/arm/emit.mlp:
```diff
@@ -56,44 +53,71 @@ let emit_reg = function
+    mutable offset_literals : offset_computation list;
+    mutable gotrel_literals : gotrel_literal list;
+    mutable symbol_literals : symbol_literal list;
+    (* Total space (in words) occupied by pending literals *)
```
The comment should probably mention that `size_literals` also takes into account the float literals, which are in the shared part of the environment.

lthls commented on asmcomp/arm/emit.mlp:
```diff
@@ -274,87 +280,72 @@ let output_epilogue f =
+let float_literal env fl =
   try
-    List.assoc f !float_literals
+    let x = List.find (fun x -> compare x.fl fl = 0) env.float_literals in
```
See my comment int the arm64 version on float comparisons.

lthls commented on asmcomp/arm/emit.mlp:
```diff
@@ -274,87 +280,72 @@ let output_epilogue f =
+  let p = env.p in
   try
-    List.assoc s !symbol_literals
+    let sl = List.find (fun x -> compare x.sym sym = 0) p.symbol_literals in
```
I would like `String.equal x.sym sym` better.

lthls commented on asmcomp/arm/emit.mlp:
```diff
@@ -982,10 +960,10 @@ let fundecl fundecl =
-  List.iter emit_call_gc !call_gc_sites;
-  List.iter emit_call_bound_error !bound_error_sites;
+  emit_all env 0 true fundecl.fun_body;
+  emit_literals env;(* CR: shouldn't it be after emit_call_gc like in arm64? *)
```
As far as I can tell, it's important that the literals are emitted right after the code, as they need to be kept fairly close to the code that uses them (see `emit_all` for details). The constraint probably doesn't apply to arm64, so the literals can safely be emitted after all code parts.

gretay-js replied:
Thanks for the explanation.

lthls commented on asmcomp/amd64/emit.mlp:
```diff
@@ -196,31 +186,21 @@ let reg64 = function
-
-let arg i n = reg i.arg.(n)
-
+let arg env i n = reg env i.arg.(n)
```
Would you mind keeping the definition of `res` here, for consistency (although it's not used outside of `emit_instr`) ?

lthls commented on asmcomp/amd64/emit.mlp:
```diff
@@ -536,30 +500,30 @@ let emit_instr fallthrough i =
           end
         end
-      end
+      end;
```
Extra (harmless) semicolon here

lthls commented on asmcomp/amd64/emit.mlp:
```diff
@@ -571,7 +535,7 @@ let emit_instr fallthrough i =
         end
       end else begin
-        emit_call func
+        emit_call func;
```
Extra (harmless) semicolon here

lthls reviewed (commented):
I've gone through all the code, and it looks very good, though I've left a few comments in places that could be cleaned up (like remaining CR comments).

There is one thing I'm not very comfortable with though, and that is the `p` field of the environment that holds data specific to each architecture.
I think it wouldn't be particularly problematic to have all those special cases directly in `Emitenv` (as an example, the `int_literals` field is specific to s390x, but is already in `Emitenv`). What do you think about it ?
It would be particularly nice if you could also add comments in `emitenv.mli` explaining which emitters use which features (all architectures, all expect this one, only for this one, ...).

gretay-js committed:
Address review comments

gretay-js committed:
Remove target-specific field of Emitenv

Put special cases directly in Emitenv and add comments
explaining which emitters use which features.

gretay-js commented:
Thank you for the review! 
I've rebased and addressed all the comments. 
There is a separate commit to remove the target-specific field of Emitenv and add all the specials case fields over. Some of these fields can be reused (gc_call related fields in particular), but not trivially, so it would be a separate PR.

gretay-js committed:
Move the definition of res next to arg (amd64)

gretay-js committed:
Update comments in Emitenv

lthls commented on asmcomp/arm64/emit.mlp:
```diff
@@ -339,33 +307,30 @@ let emit_cmpimm rs n =
   try
-    List.assoc f !float_literals
+    let eq x =
+      Float.equal (Int64.float_of_bits x.fl) (Int64.float_of_bits fl)
```
Sorry, I failed to notice that the floats were already stored as `int64`s, so the previous code (and the original `List.assoc` code) was correct. The whole point is to avoid using float comparison in the compiler, so this new code is wrong.
For clarity, I would still prefer `List.find (fun x -> Int64.equal x.fl fl) env.float_literals` rather than your original `compare`.
(If you follow this suggestion, please update `arm/emit.mlp` in the same way).

gretay-js replied:
Fixed in both arm and arm64.
Thank you for noticing this, it was my mistake (I misunderstood your comment to mean that Float comparison should be used instead of Int64).

lthls commented:
Thanks for the update. It turns out my review comment on float comparisons was wrong, so you'll need to revert that part, but apart from that everything looks good.

gretay-js committed:
Search through float_literals using Int64.equal (arm,arm64)

lthls reviewed (approved):
Looks good to me. I'll merge when CI is green.

gretay-js commented:
Thank you! 
There are so many commits, it should probably be a squash-merge, if possible.

gasche commented:
Note: I think that this merge broke `dune build @libs` (which I use for nice Merlin stuff) in trunk.

```
File "asmcomp/emitaux.mli", line 88, characters 31-55:
Error: Unbound module Emitenv
```

(I think it is related to this PR because it introduced `asmcomp/emitenv.mli`, a mli-only module)

I'm not sure how to fix the build. (cc @trefis maybe?)

lthls commented:
I think you need to add `emitenv` and `cmm_invariants` to the list of modules in the rule for `ocamloptcomp`, and also add `emitenv` to the list of `modules_without_implementation` for the same rule. At least this seems to fix the build for me.

gasche commented:
Great, thanks @lthls! Case closed.
